### PR TITLE
KCWI flux calibration

### PR DESCRIPTION
**Pull request by rcooke-ast**
This PR brings the Keck/KCWI data reduction to completion, and fixes a few minor bugs (e.g. mask_abs_line was not being passed).

Tests pass:
image

I've run this on the `Cooked` dev suite, and generated a new Cooked file...

### CODE REVIEW

File: pypeit/core/coadd.py
Comment (jhennawi) [inline pypeit/core/coadd.py]: Formally should this not be a boolean operation rather than *=? I'm sure this does the same thing but is less clear since mask_stack is a boolean array.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py:63]: I'm not following why this is either or?

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Please add documentation.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Documentation please. Inline comments are also strongly encouraged. Right now it does not look anyone else can maintain or modify this code.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: Docs, comments.

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: I'm not following why this is counts per second per angstrom and not just counts per second. Where have you divided by the wavelength grid?

File: pypeit/core/datacube.py
Comment (jhennawi) [inline pypeit/core/datacube.py]: please clean up commented regions if you no longer need them.

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py:66]: Can you add the units of a?

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py:69]: If I understand what you've done in this function teff is not a float, but rather an astropy Quantity. Please avoid making satrapy quantities inputs to functions like this. I would make the inputs floats and then clearly specify what their units are in the docs. We have a convention in PypeIt which is that we don't use astropy Quantities except internally in functions and then we also pass back floats (i.e. .value) and not quantities. This is done for very good reasons, namely computations with astropy Quantities can slow down floating point operations by a factor of 10 or more. My suggestion is thus that you pass in floats and pass out floats, and only use the quantities and units internally. I love this functionality in astropy, but I think this is the best approach given how slow it is.

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py]: maybe define 1.0e-23 with a global constant similar to PYPEIT_FLUX_SCALE?

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py]: How about we rename this counts2Nlam or something like that.

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py:794]: Please document the arrays Nlam_star, Nlam_ivar_star, and gpm_star. And also specify their units.

File: pypeit/core/flux_calib.py
Comment (jhennawi) [inline pypeit/core/flux_calib.py:818]: good pixel mask for zerpoint

File: pypeit/par/pypeitpar.py
Comment (jhennawi) [inline pypeit/par/pypeitpar.py:648]: Can you add to these docs that this functionality is currently only used for IFUs?

**Review by jhennawi — CHANGES_REQUESTED** [review]
Looks good Ryan. Great work here. My comments are relatively minor and mainly relate to documentation. Dev-suite and unit tests definitely need to be run here.

File: pypeit/core/coadd.py
Comment (rcooke-ast) [reply pypeit/core/coadd.py]: OK - done

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py:63]: It depends whether the cube is flux calibrated or not. The combined standard star datacubes are given in counts.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Done!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Too right - both implemented!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Done!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: I've expanded the comment. We are integrating over the spatial coordinates, and each pixel is a surface brightness. The input is already per Angstroms.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Thanks for pointing this out - I've fixed this, and a couple more throughout...

File: pypeit/core/flux_calib.py
Comment (rcooke-ast) [reply pypeit/core/flux_calib.py:66]: Yep - it's dimensionless  :-)

File: pypeit/core/flux_calib.py
Comment (rcooke-ast) [reply pypeit/core/flux_calib.py:69]: That's a really good point - simple to update, and done!

File: pypeit/core/flux_calib.py
Comment (rcooke-ast) [reply pypeit/core/flux_calib.py]: Sure - I'm not sure how useful this will be for other routines, as it's just how the table data are stored.

File: pypeit/core/flux_calib.py
Comment (rcooke-ast) [reply pypeit/core/flux_calib.py]: Sounds good to me - done!

File: pypeit/core/flux_calib.py
Comment (rcooke-ast) [reply pypeit/core/flux_calib.py:794]: Done

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [reply pypeit/par/pypeitpar.py:648]: I have added this, but I think @profxj and @marlageha may be using this to account for vignetting of DEIMOS (but I could be mistaken).

**Review by rcooke-ast — COMMENTED** [review]
Thanks @jhennawi for the quick and solid review!  Comments addressed in all cases. Let me know if you have anything else to add.

File: pypeit/core/datacube.py
Comment (jhennawi) [reply pypeit/core/datacube.py:63]: Okay, that is fine but then counts are not counts/s/ang unless you have divided by the exposure time (easy) and the delta_lambda (more complicated since in general it varies with pixel). So has that been done here (perhaps I missed it) or is the label of the units incorrect?

**Review by jhennawi — APPROVED** [review]
Hi @rcooke-ast, 

Thanks for addressing everything. Given that you modified the core fluxing routines the full dev suite needs to be run here. If you are not able to do that yourself, another developer can help you run in the cloud.  I'm approving, but please don't merge until the full dev suite passes.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py:63]: Thanks for checking - the units label is correct here. The exposure time and delta wave is done around line 1200. The delta wave is done on a per pixel basis (as you mention, it's not constant for every pixel, and this is accounted for).

File: pypeit/core/datacube.py
Comment (profxj) [inline pypeit/core/datacube.py:397]: Maybe note that the rebinning adopts the `mean`

File: pypeit/core/datacube.py
Comment (profxj) [inline pypeit/core/datacube.py]: while this is a pretty simple 4 lines of code, I wonder whether it should be a method somewhere.  Just consider it

File: pypeit/core/datacube.py
Comment (profxj) [inline pypeit/core/datacube.py]: Similar to the above.  If this is a generally useful method, consider having it stand-alone for wider usage.

File: pypeit/core/datacube.py
Comment (profxj) [inline pypeit/core/datacube.py]: Remove?

File: pypeit/flatfield.py
Comment (profxj) [inline pypeit/flatfield.py:1287]: is this optional?  just checking.

**Review by profxj — APPROVED** [review]
Whoa quite the PR!

Is there going to be a test of the fluxing in the DevSuite?
Am hoping so.   But am approving now anyhow.

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py:397]: Done!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Done!

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py:1287]: Yep - added a note

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: OK - I've made this its own routine, and beefed it up a little to make it more general and useful...

File: pypeit/core/datacube.py
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Done!

**Review by rcooke-ast — COMMENTED** [review]
Thanks @profxj - comments addressed. Just running some final checks before we run the dev suite

**Comment by rcooke-ast** [comment]
> Whoa quite the PR!
> 
> Is there going to be a test of the fluxing in the DevSuite? Am hoping so. But am approving now anyhow.

Thanks @profxj - I've added a test of the flux calibration in the unit tests. Hope that's OK...

**Comment by rcooke-ast** [comment]
Tests pass...
image

**Comment by rcooke-ast** [comment]
Tests and devsuite pass... merging!

image

image
(the three dev suite failures are expected failures that are fixed with an update to the dev suite)

**Pull request merged**
